src/main/java/com/google/devtools/build/lib/skyframe/BzlLoadFunction.java

```diff
@@ -972,6 +979,33 @@ private static RepositoryMapping getRepositoryMapping(
     return repositoryMappingValue.getRepositoryMapping();
   }
 
+  @Nullable
+  private static Optional<RepositoryMapping> getMainRepositoryMapping(
+      BzlLoadValue.Key key, StarlarkSemantics starlarkSemantics, Environment env)
+      throws InterruptedException {
+    if (!starlarkSemantics.getBool(BuildLanguageOptions.ENABLE_BZLMOD)) {
+      return Optional.empty();
```

Comment:
I don't understand why return empty here; you can have a repo map in legacy WORKSPACE mode.

Reply:
This has always been broken in numerous ways, @Wyverald can probably say more about it. As far as I know you also couldn't define a non-trivial repo mapping *for the main repo*, only for other repos.

Reply:
One undesirable result of a null main repo map though is that in legacy WORKSPACE mode, all non-main-repo labels will get printed in `@@` form.

Reply:
Observe that the weird corner cases for legacy WORKSPACE, bazel_tools etc. are already well-handled by `getRepositoryMapping`. So how about the following implementation of `getMainRepositoryMapping`, which falls back to `getRepositoryMapping` for the edge case logic:

```java
  @Nullable
  private static RepositoryMapping getMainRepositoryMapping(
      BzlLoadValue.Key key, StarlarkSemantics starlarkSemantics, Environment env)
      throws InterruptedException {
    boolean bzlmod = starlarkSemantics.getBool(BuildLanguageOptions.ENABLE_BZLMOD);
    RepositoryMappingValue.Key repoMappingKey;
    if (key instanceof BzlLoadValue.KeyForBuild) {
      repoMappingKey = RepositoryMappingValue.key(RepositoryName.MAIN);
    } else if ((key instanceof BzlLoadValue.KeyForBzlmod
            && !(key instanceof BzlLoadValue.KeyForBzlmodBootstrap))
        || (bzlmod && key instanceof BzlLoadValue.KeyForWorkspace)) {
      // Since the main repo mapping requires evaluating WORKSPACE, but WORKSPACE can load from
      // extension repos, requesting the full main repo mapping would cause a cycle.
      repoMappingKey = RepositoryMappingValue.KEY_FOR_ROOT_MODULE_WITHOUT_WORKSPACE_REPOS;
    } else {
      // For builtins, @bazel_tools, and legacy WORKSPACE, the key's local repo mapping can be used
      // as the main repo mapping.
      return getRepositoryMapping(key, starlarkSemantics, env);
    }
    RepositoryMappingValue mainRepositoryMappingValue =
        (RepositoryMappingValue) env.getValue(repoMappingKey);
    if (mainRepositoryMappingValue == null) {
      return null;
    }
    return mainRepositoryMappingValue.getRepositoryMapping();
  }
```

Note that this will also let you get rid of the Optional wrapper as a side effect

Reply:
Also, @brandjon - do you have any comments on putting the main repo mapping in module context? Or maybe we can rather stick it into StarlarkThread somehow (I don't see any clean way to do it, but maybe you do)?

Reply:
Thanks for the suggestion, this looks much better. I also added a test for the `--noenable_bzlmod` case.

Reply:
I talked to @brandjon about this PR in our daily sync; he suggested we try to put the main repository mapping pointer in BazelStarlarkContext rather than in module context (mainly for memory usage reasons; there may be a *vast* number of module context objects in flight in memory at any one time, but only a small number of starlark threads).

Reply:
`BazelStarlarkContext` isn't available when a module extension is evaluated. Would it be okay to introduce a new `Phase` for that?

Reply:
I am now using `BazelStarlarkContext` and the existing `WORKSPACE` phase for module extension evaluation, PTAL.

@Wyverald Could you verify that adding `BazelStarlarkContext` to module extensions' Starlark threads doesn't give extensions access to functions they shouldn't have access to?